File: 프로그래머스/lv3/43162. 네트워크/네트워크.py
def solution(n, computers):
    answer = 0
    visit = [0] * n
    
    def dfs(j):
        for idx, network in enumerate(computers[j]):
            # print(f'j: {j}, idx: {idx}, network: {network}, visit[idx]: {visit[idx]}')
            if network == 1 and visit[idx] == 0:
                visit[idx] = 1
                dfs(idx)
                
        return
    for i in range(n):
        isFind = False
        
        for j in range(n):
            if i != j:
                # 연결되었다면
                # print(f'i: {i}, j: {j} computers: {computers[i][j]}')
                if computers[i][j] == 1:
                    isFind = True
                    if visit[j] == 0:
                        # print(f'{i}와 {j}가 연결되었습니다.')
                        # 연결횟수 추가
                        answer += 1
                        visit[i] = 1
                        visit[j] = 1
                        dfs(j)
        if not isFind:
            # print(f'{i}는 네트워크 스스로 연결합니다')
            answer += 1
    return answer

File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;
typedef vector<vector<int>> vvi;

void dfs(int start, int n, vvi & computers, vector<int> & visited) {
    visited[start] = 1;
    
    for (int end = 0; end < n; end++) { 
        if (start == end) continue;
        
        if (computers[start][end] == 1 && !visited[end]) {
            dfs(end, n, computers, visited);
        }
    }    
} 
    
int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    vector<int> visited(n + 1, 0);
    
    for (int start = 0; start < n; start++) {
        if (!visited[start]) {
            answer++;
            dfs(start, n, computers, visited);
        }
    }
    
    return answer;
}